### PR TITLE
expose options.payload in index.js for issue #219

### PR DESCRIPTION
**Pull request by hfwittmann**
Exposes the payload authentication functionality.

Can be used to introduce acl type authorisation.

example:

```javascript
options.payload = (request, reply) => {

    const hasPermission = internals.rolePermissions(request.auth.credentials.scope, request.payload.entity);

    if (hasPermission) {
        return reply.continue();
    }

    return reply(Boom.unauthorized('Scope has no permission to do this'));
};

```

In this example internals.rolePermissions is a function that needs to be defined. It decides whether the used tokens scope has the permission to e.g. edit the payload's entity

### CODE REVIEW

**Comment by hfwittmann** [comment]
I think it would be very useful to have this in there. No interest?

**Comment by nelsonic** [comment]
Hi @hfwittmann! it's not a question of "No interest" we just need some clarity round the _use-case_.
the `options` object is defined when you _register_ the plugin (_during server startup_) so technically you _already_ have access to the value/object you have set in `options.payload` within the context of your application. 

If you want to use the `payload` that the client/user sent in their request that's _also_ already available on the `request` object in your route handler. 

Please help us _understand_ the context for using `options.payload`, thanks! 👍

**Comment by hfwittmann** [comment]
Thanks for your reply.

Actually, what you say about the payload being available in the request object is not true in the validation callback. This was discussed here:
https://github.com/dwyl/hapi-auth-jwt2/issues/219

Therefore, currently, in my view, you can't use payload dependent authorization with dwyl/hapi-auth-jwt2

Felix

**Comment by nelsonic** [comment]
@hfwittmann you are _correct_ in that issue #219 mentions `request.payload`
_however_ it does not mention `options.payload` ... 
_again_, `options` is defined when the plugin is instantiated and exposing `options.payload` will _not_ make the `request.payload` available in your route's context.

Perhaps the _best_ way for you to _illustrate_ your use-case is to create a _Test_ showing _exactly_ how you expect/intent to use this functionality ... 🤔

**Comment by hfwittmann** [comment]
Thanks for your reply.

Just to be sure we're talking about the same thing please have a look at Route configuration in
// https://hapijs.com/tutorials/auth?lang=en_US#route-configuration

I have prepared a couple of files to showcase the issue.

Steps:

- Unzip archive into a directory
- npm install
- nodemon simpleHapi.js or alternatively nodemon simpleHapi.payloadAuth.js

**Please note that nodemon simpleHapi.payloadAuth.js wont run unless my proposed change (i.e, payload: options.payload, in line 212) is made.**

after nodemon simpleHapi.js / nodemon simpleHapi.payloadAuth.js you can go to http://localhost:3000/documentation where the routes are available via swagger. Alternatively you can run the curl type commands at the end of the files to see what's going on.





[payloadDrivenAuthentication.zip](https://github.com/dwyl/hapi-auth-jwt2/files/1183383/payloadDrivenAuthentication.zip)

**Comment by nelsonic** [comment]
@hfwittmann, no offence, but, you "lost me" at "_unzip archive_" ... 😕  
(_there's a [simpler](https://youtu.be/L_o_O7v1ews) way to share code ..._ 🙄 
_sharing a zip file/folder containing code on GitHub is like serving Vodka at an "AA" meeting..._😉 )

_Please_ create an _automated test_ in your Fork and commit it so that Travis-CI runs the test.

> There are _many_ example tests you can draw inspiration from: 
https://github.com/dwyl/hapi-auth-jwt2/tree/master/test
Every "_new feature_" requires a test else it risks being "broken" in a "re-factor".

**Comment by hfwittmann** [comment]
No offence taken, tests are in.

**Comment by hfwittmann** [comment]
Anything I can do?

File: test/options_payload_validation.js [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+                reply({ text: 'You used a Token!', payload: request.payload })
+                    .header("Authorization", request.headers.authorization);
+            }
+        }
+    ]);
+
+});
+
+
+test('1', function (t) {
```
Comment (nelsonic) [inline test/options_payload_validation.js]: there are a couple of _duplicate_ (_redundant_) tests in this file, would you mind if I streamlined it?

**Pull request merged**